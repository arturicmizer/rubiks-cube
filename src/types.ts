export enum Color {
    LAPTEUH = 1,
    YELLOW,
    BLUE,
    RED,
    GWYNNNPLAINE,
    ORANGE
}


export type SideName = 'left' | 'right' | 'down' | 'up' | 'front' | 'back';

export interface Cell {
    color: Color,
    id: string;
}

export type CubeRow = readonly [Cell, Cell, Cell];
export  type CubeSide = readonly [
    CubeRow,
    CubeRow,
    CubeRow,
]

export type CubeData = [
    CubeSide,
    CubeSide,
    CubeSide,
    CubeSide,
    CubeSide,
    CubeSide,
]

export type Cycle = readonly [number, number, number, number];

type Operation = -1 | 0 | 1;
export type ShiftOperation = readonly [
    readonly [Operation, Operation, Operation],
    readonly [Operation, Operation, Operation],
    readonly [Operation, Operation, Operation],
];

export type RotateFn = (s: CubeSide) => CubeSide;
export type RotateMatrix = readonly [RotateFn, RotateFn, RotateFn, RotateFn];


export type AnimationDirection = 'left' | 'top' | 'right' | 'bottom' | undefined;
export type AnimationDirections = (AnimationDirection)[][];

export interface UpdateAction {
    operation: ShiftOperation;
    cycle: Cycle;
    rotateMatrix?: RotateMatrix;
    animations: [AnimationDirection, AnimationDirection];
}

