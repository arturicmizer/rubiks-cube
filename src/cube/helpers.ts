import {Cell, Color, CubeData, CubeRow, CubeSide, Cycle, RotateFn, RotateMatrix, ShiftOperation} from "../types";


export function createSide(color: Color, startIndex: number): CubeSide {
    const result: Partial<CubeRow[]> = []
    for (let i = 0; i < 3; i++) {
        const row: Cell[] = [];
        for (let j = 0; j < 3; j++) {
            row.push({
                color: color,
                id: (startIndex++).toString()
            })
        }

        result.push(row as unknown as CubeRow);
    }

    // TOOD(kirjs): Figure out if there's a good way to type it.
    return result as unknown as CubeSide;
}

export function createCube(): CubeData {
    const result: CubeSide[] = [];
    let i = 0;
    for (const color of Object.values(Color)) {
        if (typeof color === 'number') {
            result.push(createSide(color, (i++) * 9));
        }
    }

    return result as CubeData;
}


export function shift(cube: CubeData, cycle: Cycle, operation: ShiftOperation): CubeData {
    return cube.map((side, i) => {
        let index = cycle.indexOf(i);
        if (index === -1) {
            return side;
        } else {

            return side.map((row, rowIndex) => {
                return row.map((cell, cellIndex) => {
                    const o = operation[rowIndex][cellIndex];
                    const nextIndex = (index + o + cycle.length) % cycle.length;
                    return cube[cycle[nextIndex]][rowIndex][cellIndex];
                });
            })
        }
    }) as CubeData;
}


export const rotate90 = (side: CubeSide): CubeSide => {
    return side.map((row, i) =>
        row.map((val, j) => side[side.length - 1 - j][i])
    ) as unknown as CubeSide;
};

export const rotate270 = (side: CubeSide): CubeSide => {
    return rotate90(rotate90(rotate90(side)))
};

export const rotate180 = (side: CubeSide): CubeSide => {
    return rotate90(rotate90(side))
};

export const rotate0 = (side: CubeSide): CubeSide => {
    return side;
};


const reverseRotateMap = new Map<RotateFn, RotateFn>();
reverseRotateMap.set(rotate270, rotate90);
reverseRotateMap.set(rotate180, rotate180);
reverseRotateMap.set(rotate90, rotate270);
reverseRotateMap.set(rotate0, rotate0);

export function rotateCube(cube: CubeData, cycle: Cycle, matrix: RotateMatrix): CubeData {
    return cube.map((side, i) => {
        const index = cycle.indexOf(i);
        if (index === -1) {
            return side;
        } else {
            return matrix[index](side);
        }
    }) as CubeData;
}

export function reverseRotateMatrix(matrix: RotateMatrix) {
    return matrix.map(r => reverseRotateMap.get(r) as RotateFn) as unknown as RotateMatrix;
}