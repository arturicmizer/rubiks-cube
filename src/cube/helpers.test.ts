import {createCube, rotate90, shift} from "./helpers";
import {CubeSide, ShiftOperation} from "../types";

describe('shift', function () {
    it('lol', function () {
        const cube = createCube();
        const cycle = [2, 3, 4, 5] as const;

        const operation: ShiftOperation = [
            [1, 1, 1],
            [0, 0, 0],
            [0, 0, 0],
        ];

        expect(shift(cube, cycle, operation)).toMatchSnapshot();
    });
});
describe('rotate', function () {
    it('rotates', function () {
        const side = [
            [1, 2, 3],
            [4, 5, 6],
            [5, 5, 5],
        ] as CubeSide;


        expect(rotate90(side)).toEqual([
            [5, 4, 1],
            [5, 5, 2],
            [5, 6, 3]]);
    });
});