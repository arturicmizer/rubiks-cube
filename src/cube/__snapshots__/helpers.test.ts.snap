// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shift lol 1`] = `
Array [
  Array [
    Array [
      1,
      1,
      1,
    ],
    Array [
      1,
      1,
      1,
    ],
    Array [
      1,
      1,
      1,
    ],
  ],
  Array [
    Array [
      2,
      2,
      2,
    ],
    Array [
      2,
      2,
      2,
    ],
    Array [
      2,
      2,
      2,
    ],
  ],
  Array [
    Array [
      4,
      4,
      4,
    ],
    Array [
      3,
      3,
      3,
    ],
    Array [
      3,
      3,
      3,
    ],
  ],
  Array [
    Array [
      5,
      5,
      5,
    ],
    Array [
      4,
      4,
      4,
    ],
    Array [
      4,
      4,
      4,
    ],
  ],
  Array [
    Array [
      6,
      6,
      6,
    ],
    Array [
      5,
      5,
      5,
    ],
    Array [
      5,
      5,
      5,
    ],
  ],
  Array [
    Array [
      3,
      3,
      3,
    ],
    Array [
      6,
      6,
      6,
    ],
    Array [
      6,
      6,
      6,
    ],
  ],
]
`;
